tensorflow_probability/python/distributions/empirical.py

```diff
@@ -0,0 +1,204 @@
+    return tf.constant([], dtype=tf.int32)
+
+  def _batch_shape(self):
+    return tf.TensorShape([])
+
+  def _event_shape_tensor(self):
+    return tf.shape(self.params)[1:]
+
+  def _event_shape(self):
+    return self.params.shape[1:]
+
+  def _mean(self):
+    return tf.reduce_mean(self.params, 0)
+
+  def _mode(self):
+    params = array_ops.reshape(self.params, [-1])
+    unique = tf.unique_with_counts(params)
+    return tf.gather(unique.y, tf.argmax(unique.count))
+
+  def _stddev(self):
+    # broadcasting n x shape - shape = n x shape
+    r = self.params - self.mean()
+    var = tf.cast(tf.reduce_mean(tf.square(r), 0), tf.float32)
+    return tf.sqrt(var)
+
+  def _variance(self):
+    return tf.square(self.stddev())
+
+  def _sample_n(self, n, seed=None):
+    input_tensor = self.params
+    if len(input_tensor.shape) == 0:
+      input_tensor = tf.expand_dims(input_tensor, 0)
+      multiples = tf.concat(
+          [tf.expand_dims(n, 0), [1] * len(self.event_shape)], 0)
+      return tf.tile(input_tensor, multiples)
+    else:
+      probs = tf.ones([self.n_sample]) / tf.cast(self.n_sample, dtype=tf.float32)
+      cat = Categorical(probs)
+      indices = cat._sample_n(n, seed)
+      tensor = tf.gather(input_tensor, indices)
+      return tensor
+
+  def _log_prob(self, event):
+    params = tf.convert_to_tensor(self.params)
+    labels = tf.convert_to_tensor(event)
+    if self.validate_args:
+      labels = util.embed_check_integer_casting_closed(
+        labels, target_dtype=tf.bool)
+
+    # Raise an error when the labels shape does not equal the event shape.
+    # Note: This could break users who call this with bad labels, but
+    # disregard the bad results.
+    labels_static_shape = labels.get_shape()
+    labels_shape = array_ops.shape(labels)
+    static_shapes_fully_defined = (
+            labels_static_shape.is_fully_defined() and
+            self.params.get_shape()[1:].is_fully_defined())
+    if params.get_shape().ndims is not None and (
+            labels_static_shape.ndims is not None and
+            labels_static_shape.ndims != params.get_shape().ndims - 1):
+      raise ValueError("Rank mismatch: Rank of labels (received {}) should "
+                        "equal rank of params minus 1 (received {})." \
+                       .format(labels_static_shape.ndims, params.get_shape().ndims))
+
+    if (static_shapes_fully_defined and
+              labels_static_shape != params.get_shape()[1:]):
+      raise ValueError("Shape mismatch: The shape of labels (received {}) "
+                        "should equal the event shape (received {})." \
+                        .format(labels_static_shape, params.get_shape()[1:]))
+
+    # Reshape logits to 2 dim, labels to 1 dim.
+    params = array_ops.reshape(params, [self.n_sample, -1])
+    labels = array_ops.reshape(labels, [-1])
+
+    elements = tf.cast(tf.equal(params, labels), tf.int32)
+    total_count = tf.reduce_sum(tf.reduce_mean(elements, 1))
```

Comment:
Are you sure `reduce_mean` is the right operation here? It seems like you'd want `reduce_prod` so that a sample only counts towards the log prob if it actually represents the same point in n-d space, i.e., matches the provided event in every dimension. 
(if so, and this is a bug, then you should also add a test that would have caught this bug).

Reply:
Agree and Fixed. `reduce_prod` will be the right operation to serve for row-wise comparison here. I added several tests on this.